src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -2080,4 +2082,44 @@ protected void notifyShutdown() {
         // No-op, handled in the suspendAll terminator
     }
 
+    // TODO: Core could use a QuietDownListener class with `onQuietDown` and `onCancelQuietDown` methods.
+    @Restricted(NoExternalUse.class)
+    @Extension
+    public static class QuietDownListener extends PeriodicWork {
```

Comment:
You should not need a global extension to do this. Rather, when you `saveProgramIfPossible(true)` inside `scheduleRun()`, use `Timer` to set up a task which checks (say) every second to see if _this_ program can resume. (If you use an anonymous inner `Runnable` class, you can `Timer.get().submit(this)` to reschedule if Jenkins is still quieting down.)

Reply:
Ok, I'll look into it. IIUC, you mean that if a build is executing something like a long-running `sh` step when quiet mode is enabled, there is no need to call `scheduleRun` on it if quiet mode is canceled _unless_ the `sh` step completed while quiet mode was enabled, is that right?

Reply:
No, that is not what I meant at all. Let me try to rephrase. You are fixing what I consider to be a corner case that many people will never hit. Yet the implementation registers a global extension that does something in the JVM every minute, even if no Pipeline builds are running, and needs to iterate all running Pipeline builds. Instead you can just create a timer task in a _particular_ build that you _know_ was suspended due to quiet mode because we just ran https://github.com/jenkinsci/workflow-cps-plugin/blob/c4ecc1d33d66449531332072db8387347e35c5ba/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java#L274-L278 which will incur that overhead if only and if a build actually needs it, and there is no need to look up `FlowExecutionList` etc. because we have references to all the relevant objects right here.

Reply:
My confusion is that it doesn't seem like there is a meaningful difference between a global extension that only loops over all Pipeline builds when quiet mode is canceled and running a timer task per suspended Pipeline build while quiet mode is enabled, unless there is some reason to expect that the set of suspended Pipeline builds will normally be smaller than the set of all Pipeline builds, so I was trying to figure out if that was the case.

Reply:
> there is some reason to expect that the set of suspended Pipeline builds will normally be smaller than the set of all Pipeline builds

Consider the normal case that Jenkins has not been set to quiet down in the current session at all. The current patch would be waking up every minute. I was merely proposing to activate the code in this patch only if and when a particular Pipeline build was actually suspended due to Jenkins quieting down.

Reply:
…and scheduling a task which was actually specific to that build, independent of all others, which ought to make it more obvious from reading the code that it is doing the right thing.

Reply:
> I was merely proposing to activate the code in this patch only if and when a particular Pipeline build was actually suspended due to Jenkins quieting down.

Sure, but then instead of waking up Jenkins once a minute (and usually doing more or less nothing), you are waking up Jenkins even more frequently than once a minute for every Pipeline build when quiet mode is enabled (and usually doing more or less nothing). Depending on how often you enable quiet mode and how many builds are running concurrently, I could see either approach being preferable, so I just wanted to check if I was missing something important that might tip the balance.

Regardless, I switched to your approach in e026365, so let me know what you think.